src/Middleware/HttpLogging/src/HttpRequestLog.cs

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    internal sealed class HttpRequestLog : IReadOnlyList<KeyValuePair<string, object?>>
+    {
+        private readonly List<KeyValuePair<string, object?>> _keyValues;
+        private string? _cachedToString;
+
+        internal static readonly Func<object, Exception?, string> Callback = (state, exception) => ((HttpRequestLog)state).ToString();
+
+        public HttpRequestLog(List<KeyValuePair<string, object?>> keyValues)
+        {
+            _keyValues = keyValues;
+        }
+
+        public KeyValuePair<string, object?> this[int index] => _keyValues[index];
+
+        public int Count => _keyValues.Count;
+
+        public IEnumerator<KeyValuePair<string, object?>> GetEnumerator()
+        {
+            var count = _keyValues.Count;
+            for (var i = 0; i < count; i++)
```

Comment:
```suggestion
            foreach (var kvp in CollectionsMarshal.AsSpan(_keyValues))
```

Reply:
That is giving me the error: CS8344: foreach statement cannot operate on enumerators of type 'Span<KeyValuePair<string, string>>.Enumerator' in async or iterator methods because 'Span<KeyValuePair<string, string>>.Enumerator' is a ref struct.

Reply:
Yea you can't use span in a method that yields the stack.